File: backend/internal/domain/purchase_order.go
package domain

import "time"

// PurchaseOrder tracks goods ordered from a supplier.
type PurchaseOrder struct {
	ID                string     `db:"id"`
	StoreID           string     `db:"store_id"`
	SupplierID        *string    `db:"supplier_id"`
	PONumber          string     `db:"po_number"`
	Status            string     `db:"status"` // draft | ordered | received | canceled
	TotalAmount       float64    `db:"total_amount"`
	OrderedBy         string     `db:"ordered_by"`
	ReceivedBy        *string    `db:"received_by"`
	OrderedAt         *time.Time `db:"ordered_at"`
	ReceivedAt        *time.Time `db:"received_at"`
	Notes             *string    `db:"notes"`
	CreatedAt         time.Time  `db:"created_at"`
	UpdatedAt         time.Time  `db:"updated_at"`
	SupplierSignature *string    `db:"supplier_signature"`
	BuyerSignature    *string    `db:"buyer_signature"`
	SupplierSignedAt  *time.Time `db:"supplier_signed_at"`
	BuyerSignedAt     *time.Time `db:"buyer_signed_at"`

	// Populated via JOIN
	SupplierName   *string  `db:"supplier_name"`
	OrderedByName  string   `db:"ordered_by_name"`
	ReceivedByName *string  `db:"received_by_name"`
	TotalItems     int      `db:"total_items"`
	Items          []POItem `db:"-"`

	// Payment aggregation (loaded separately)
	AmountPaid    float64    `db:"amount_paid"`
	AmountDue     float64    `db:"amount_due"`
	NextDeadline  *time.Time `db:"next_deadline"`
	PaymentStatus string     `db:"payment_status"` // unpaid | partial | paid
}

// POItem is a single line in a purchase order.
type POItem struct {
	ID          string  `db:"id"`
	POID        string  `db:"po_id"`
	ProductID   string  `db:"product_id"`
	Quantity    float64 `db:"quantity"`
	UnitCost    float64 `db:"unit_cost"`
	ReceivedQty float64 `db:"received_qty"`
	Subtotal    float64 `db:"subtotal"`

	// Populated via JOIN
	ProductName string `db:"product_name"`
	ProductSKU  string `db:"product_sku"`
	Unit        string `db:"unit"`
}

// POPayment records a simple payment against a purchase order (legacy).
type POPayment struct {
	ID         string    `db:"id"`
	POID       string    `db:"po_id"`
	StoreID    string    `db:"store_id"`
	Amount     float64   `db:"amount"`
	Note       *string   `db:"note"`
	PaidBy     string    `db:"paid_by"`
	PaidByName string    `db:"paid_by_name"`
	PaidAt     time.Time `db:"paid_at"`
}

// ─── Termin (Installment) System ───────────────────────────────────────────────────

// POTermin is one installment in a purchase order payment schedule.
// status transitions: unpaid → partial → paid (overdue is a derived view).
type POTermin struct {
	ID           string    `db:"id"`
	POID         string    `db:"po_id"`
	TerminNumber int       `db:"termin_number"`
	Amount       float64   `db:"amount"`
	DueDate      time.Time `db:"due_date"`
	Status       string    `db:"status"` // unpaid | partial | paid | overdue
	Notes        string    `db:"notes"`
	CreatedAt    time.Time `db:"created_at"`
	UpdatedAt    time.Time `db:"updated_at"`

	// Populated via aggregation
	AmountPaid float64         `db:"amount_paid"`
	AmountDue  float64         `db:"amount_due"`
	Payments   []PaymentRecord `db:"-"`
}

// PaymentRecord is a single payment transaction against one POTermin.
type PaymentRecord struct {
	ID            string    `db:"id"`
	TerminID      string    `db:"termin_id"`
	AmountPaid    float64   `db:"amount_paid"`
	PaymentDate   time.Time `db:"payment_date"`
	PaymentMethod string    `db:"payment_method"` // cash | transfer | check | other
	Notes         string    `db:"notes"`
	RecordedBy    *string   `db:"recorded_by"`
	CreatedAt     time.Time `db:"created_at"`

	// Populated via JOIN
	RecordedByName string `db:"recorded_by_name"`
}

// PODebtSummary aggregates payment status across all termins for a PO.
type PODebtSummary struct {
	POID          string  `db:"po_id"`
	TotalAmount   float64 `db:"total_amount"`
	TotalTermin   float64 `db:"total_termin"` // sum of termin amounts
	TotalPaid     float64 `db:"total_paid"`
	RemainingDebt float64 `db:"remaining_debt"`
	Status        string  // unpaid | partial | paid
	TerminCount   int     `db:"termin_count"`
	OverdueCount  int     `db:"overdue_count"`
}

File: frontend/src/types/purchase-order.ts
export interface POItem {
  id: string;
  product_id: string;
  product_name: string;
  product_sku: string;
  unit: string;
  quantity: number;
  unit_cost: number;
  received_qty: number;
  subtotal: number;
}

export interface PurchaseOrder {
  id: string;
  store_id: string;
  supplier_id?: string;
  supplier_name?: string;
  po_number: string;
  status: 'draft' | 'ordered' | 'received' | 'cancelled';
  total_amount: number;
  total_items?: number;
  ordered_by_name: string;
  received_by_name?: string;
  ordered_at?: string;
  received_at?: string;
  next_deadline?: string;
  amount_paid?: number;
  amount_due?: number;
  payment_status?: string;
  notes: string;
  items: POItem[];
  created_at: string;
  updated_at: string;
}

export interface PaymentRecord {
  id: string;
  termin_id: string;
  amount: number;
  paid_at: string;
  paid_by: string;
  paid_by_name: string;
  note?: string;
  payment_method?: string;
  amount_paid?: number;
  payment_date?: string;
  notes?: string;
  recorded_by_name?: string;
  created_at: string;
}

export interface Termin {
  id: string;
  po_id: string;
  termin_number: number;
  amount: number;
  due_date: string;
  status: 'unpaid' | 'partial' | 'paid' | 'overdue';
  notes: string;
  amount_paid: number;
  amount_due: number;
  is_overdue: boolean;
  payments: PaymentRecord[];
  created_at: string;
}

export interface PODebtSummary {
  po_id: string;
  po_number: string;
  total_amount: number;
  total_termin: number;
  total_paid: number;
  remaining_debt: number;
  status: 'unpaid' | 'partial' | 'paid';
  termin_count: number;
  overdue_count: number;
}

export interface PODocumentData {
  doc_type: 'invoice' | 'receipt' | 'termin_agreement';
  generated_at: string;
  po: {
    id: string;
    po_number: string;
    supplier_name?: string;
    total_amount: number;
    status: string;
    notes: string;
    created_at: string;
    items?: POItem[];
  };
  debt_summary: PODebtSummary;
  termins: Termin[];
  supplier_name: string;
  store_name?: string;
}

export interface CreateTerminScheduleRequest {
  termins: {
    termin_number: number;
    amount: number;
    due_date: string;
    notes?: string;
  }[];
}

export interface RecordPaymentRequest {
  amount_paid: number;
  payment_date: string;
  payment_method: 'cash' | 'transfer' | 'check' | 'other';
  notes?: string;
}

export interface CreatePORequest {
  supplier_id?: string;
  notes?: string;
  items: Array<{ product_id: string; quantity: number; unit_cost: number }>;
}

export interface POPayment {
  id: string;
  termin_id: string;
  amount: number;
  paid_by_name: string;
  paid_at: string;
  note?: string;
}

export interface PayableSummary {
  total_debt: number;
  total_paid: number;
  total_outstanding: number;
  unpaid_count: number;
  partial_count: number;
}

export interface POListParams {
  status?: string;
  per_page?: number;
  page?: number;
  supplier_id?: string;
}
